rootfs/etc/nginx/lua/balancer/sticky_persistent.lua

```diff
@@ -0,0 +1,56 @@
+-- An affinity mode which makes sure a session is always routed to the same endpoint.
+-- The advantage of this mode is that a user will never lose his session.
+-- The drawback of this mode is that when scaling up a deployment, sessions will not
+-- be rebalanced.
+--
+local balancer_sticky = require("balancer.sticky")
+local util = require("util")
+local util_nodemap = require("util.nodemap")
+
+local _M = balancer_sticky:new()
+
+function _M.new(self, backend)
+  local nodes = util.get_nodes(backend.endpoints)
+  local hash_salt = backend["name"]
+
+  local o = {
+    name = "sticky_persistent",
+    instance = util_nodemap:new(nodes, hash_salt)
+  }
+
+  setmetatable(o, self)
+  self.__index = self
+
+  balancer_sticky.sync(o, backend)
+
+  return o
+end
+
+function _M.get_routing_key(self)
+  local cookie_value = self:get_cookie()
+
+  if cookie_value then
+    -- format <timestamp>.<workder-pid>.<routing-key>
+    local routing_key = string.match(cookie_value, '[^\\.]+$')
+
+    if routing_key == nil then
+      local err = string.format("Failed to extract routing key from cookie '%s'!", cookie_value)
+      return nil, err
+    end
+
+    return routing_key, nil
+  end
+
+  return nil, nil
+end
+
+function _M.set_routing_key(self, key)
+  local value = string.format("%s.%s.%s", ngx.now(), ngx.worker.pid(), key)
```

Comment:
Why do you need `ngx.now()` and `ngx.worker.pid()`? In `sticky_balanced` we used them to introduce some entropy for better balancing but in `sticky_persistent` I can not think of a reason to keep them.

If we can avoid this we can then just put set/get_routing_key in sticky.lua only and not deal with error handling https://github.com/kubernetes/ingress-nginx/pull/4514/files#r324818550

Reply:
They key is not designed for that kind of use. I think we should change it to hash-code only.